### PR TITLE
ISPN-10730 INADDR_ANY support for Hot Rod

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-10730

* Support INADDR_ANY in Hot Rod
* Add INADDR_ANY support to the server runtime
  * Also support multiple fallback address resolution strategies
* Document the server network and socket binding configuration

### CODE REVIEW

**Comment by tristantarrant** [comment]
@oraNod there are some server docs you should look at here

File: documentation/src/main/asciidoc/topics/ref_server_interfaces.adoc [hunk 1]
```diff
@@ -0,0 +1,130 @@
+[id='server_interfaces-{context}']
+= Network Interfaces
+{brandname} servers can use different strategies for binding to IP addresses.
+
+== Default Strategy
+Uses an `inet-address` strategy that maps a single `public` interface to the
+IPv4 loopback address (`127.0.0.1`).
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/ref_server_interfaces.adoc]: Calling it 'Address Strategy'

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      return result;
+   }
+
+   @Override
+   public int compareTo(Address o) {
+      if (o instanceof MultiHomedServerAddress) {
+         MultiHomedServerAddress oa = (MultiHomedServerAddress) o;
+         int cmp = addresses.size() - oa.addresses.size();
+         if (cmp == 0) {
+            cmp = port - oa.port;
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java]: We are also using the number of addresses. As for implementing `Comparable`, that is a requirement of the root interface `Address`. What is it used for in embedded ?

File: server/runtime/src/main/java/org/infinispan/server/configuration/InterfaceConfigurationBuilder.java [hunk 3]
```diff
@@ -18,7 +20,7 @@
       attributes.attribute(NAME).set(name);
    }
 
-   public InterfaceConfigurationBuilder address(AddressType addressType, String addressValue) {
+   public InterfaceConfigurationBuilder address(AddressType addressType, String addressValue) throws IOException {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/configuration/InterfaceConfigurationBuilder.java]: If you look at server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java#L262 I am adding all of the possible IOExceptions to a CacheConfigurationException. Wrapping it here would mean having to unwrap it there.

File: documentation/src/main/asciidoc/topics/con_server_network.adoc [hunk 4]
```diff
@@ -0,0 +1,5 @@
+= Network Configuration
+
+{brandname} servers map endpoints to interfaces and ports using the concept of `socket-bindings`.
+
+{brandname} servers use `socket-bindings` to map
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/con_server_network.adoc]: Fixed

File: documentation/src/main/asciidoc/topics/ref_server_interfaces.adoc [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+====
+
+[source,xml]
+----
+include::config_examples/server_network_matchaddress.xml[]
+----
+
+== Fallback Strategy
+Interface configurations can include multiple strategies. {brandname} servers
+try each strategy in the declared order.
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/ref_server_interfaces.adoc]: https://issues.jboss.org/browse/ISPN-11021

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+      if (port != that.port) return false;
+      return addresses.equals(that.addresses);
+
+   }
+
+//   // IMPORTANT NOTE: Hot Rod protocol agrees to this calculation for a node
+//   // address hash code calculation, so any changes to the implementation
+//   // require modification of the protocol.
+//   override def hashCode() = Arrays.hashCode(
+//      "%s:%d".format(host, port).getBytes(UTF8))
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java]: Removed

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+         if (inetAddressMatchesInterfaceAddress(localAddress.getAddress(), address.address.getAddress(), address.prefixLength)) {
+            return address.address.getHostAddress();
+         }
+      }
+      throw new IllegalArgumentException("No interface address matching " + localAddress);
+   }
+
+   static byte[] netMaskByPrefix = {(byte) 128, (byte) 192, (byte) 224, (byte) 240, (byte) 248, (byte) 252, (byte) 254};
+
+   public static String toNetMask(InterfaceAddress interfaceAddress) throws UnknownHostException {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java]: Removed

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            prefix = 0;
+         } else {
+            netmask[i] = 0;
+         }
+      }
+      return InetAddress.getByAddress(netmask).toString();
+   }
+
+   /**
+    * Checks whether the supplied network address matches the interfaceAddress
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java]: I have added a comment about using the prefix length to compare addresses

File: documentation/src/main/asciidoc/topics/proc_server_changing_bind_address.adoc [hunk 9]
```diff
@@ -0,0 +1,18 @@
+[id='changing_bind_address-{context}']
+= Changing the Default Bind Address for {brandname} Servers
+Use the `-b` switch with the {brandname} CLI or the `infinispan.bind.address`
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/proc_server_changing_bind_address.adoc]: Fixed

**Comment by oraNod** [comment]
> I added lots of small comments on the documentation part.
> 
> Mainly I think we need to be more consistent about the binding happening on an IP address, because multiple network interfaces could have the same IP address, either for higher bandwidth, high availability, or in my case to be able to use the ethernet connection while the laptop is docked without losing any ssh/vpn connection.

@danberindei hey, Dan. really good docs feedback. I couldn't add your commits directly (permissions) so I just added them as a separate commit. hope that's ok.

**Comment by tristantarrant** [comment]
Rebased and merged @oraNod 's doc changes

**Comment by gustavonalle** [comment]
I will leave this one to @danberindei as he started reviewing

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java]: I would also expect the number addresses to be the same for all the nodes in a cluster most of the time.

I believe `SyncConsistentHashFactory` needs to sort addresses to make the algorithm deterministic, so if 2 nodes have the same hash code and a merge reorders the members list we get the same CH.

But I'm not sure why `ServerAddress` extends `Address`, it's not like we can send messages to a node using its `ServerAddress`.

File: server/runtime/src/main/java/org/infinispan/server/configuration/InterfaceConfigurationBuilder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply server/runtime/src/main/java/org/infinispan/server/configuration/InterfaceConfigurationBuilder.java]: I noticed that, but I thought we'd want to wrap the exception when using the programmatic configuration as well.

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/MultiHomedServerAddress.java]: https://github.com/infinispan/infinispan/commit/b11bbe11792c96b14c3007e1304a25c361f957cf
seems to work

**Comment by tristantarrant** [comment]
@danberindei I included your changes that remove the inheritance of Address from ServerAddress

**Comment by danberindei** [comment]
Integrated, thanks Tristan!

**Pull request merged**